components/client/src/main/java/com/hotels/styx/client/HttpConfig.java

```diff
@@ -15,37 +15,31 @@
  */
 package com.hotels.styx.client;
 
-import com.fasterxml.jackson.annotation.JsonCreator;
-import com.fasterxml.jackson.annotation.JsonProperty;
-
-import static com.google.common.base.MoreObjects.firstNonNull;
 import static java.util.Arrays.asList;
 import static java.util.Collections.emptyList;
 
 /**
  * HTTP configuration settings, used to connect to origins.
  */
 public final class HttpConfig {
+
+    private static final int USE_DEFAULT_MAX_HEADER_SIZE = 0;
+    private static final int DEFAULT_MAX_HEADER_SIZE = 8192;
+
     private boolean compress;
     private final int maxInitialLength;
     private final int maxHeadersSize;
     private final int maxChunkSize;
-    private int maxContentLength = 65536;
-    private Iterable<ChannelOptionSetting> settings = emptyList();
-
-    @JsonCreator
-    HttpConfig(@JsonProperty("maxChunkSize") Integer maxChunkSize,
-               @JsonProperty("maxHeaderSize") Integer maxHeadersSize,
-               @JsonProperty("maxInitialLength") Integer maxInitialLength) {
-        this.maxChunkSize = firstNonNull(maxChunkSize, 8192);
-        this.maxHeadersSize = firstNonNull(maxHeadersSize, 8192);
-        this.maxInitialLength = firstNonNull(maxInitialLength, 4096);
-    }
+    private int maxContentLength;
+    private Iterable<ChannelOptionSetting> settings;
+
 
     private HttpConfig(Builder builder) {
         this.compress = builder.compress;
         this.maxInitialLength = builder.maxInitialLength;
-        this.maxHeadersSize = builder.maxHeadersSize;
+        this.maxHeadersSize = builder.maxHeadersSize == USE_DEFAULT_MAX_HEADER_SIZE
+                ? DEFAULT_MAX_HEADER_SIZE
+                : builder.maxHeadersSize;
```

Comment:
I don't understand this. Builder already defaults to `DEFAULT_MAX_HEADER_SIZE`. Therefore testing for zero value shouldn't be necessary?

I would then add a check to `Builder.setMaxSize` to ensure that consumers don't attempt to set it to zero:

        public Builder setMaxHeadersSize(int maxHeaderSize) {
             // Something like this
            if (maxHeaderSize <= 0) {
                throw new IllegalArgumentException("...");
            }

            this.maxHeadersSize = maxHeaderSize;
            return this;
        }

Alternatively we could do this in the build() method:

        public HttpConfig build() {
            this.maxHeaderSize = this.maxHeaderSize == 0 : DEFAULT_MAX_HEADER_SIZE: this.maxHeaderSize;
            return new HttpConfig(this);
        }

Reply:
As mentioned, 0 is the default value passed from the BackendService class so it can use an integer to represent the value. It is documented in the setter for BackendService but not in the HttpConfig one, let me add it. 
Probably we could find a better alternative...

I don't see much of a difference between reading 0 in the builder or in the constructor, but I don't know if I like modifying the contents of the builder to get the values I want in the final object. Maybe we can remove the constant for 0 though.

Reply:
This is from the `BackendService` class:

    public static final int DEFAULT_RESPONSE_TIMEOUT_MILLIS = 1000;
    public static final int USE_DEFAULT_MAX_HEADER_SIZE = 0;

This looks bit odd because the default value is declared for the response timeout, but not for the header size. 
I would make the DEFAULT_HEADER_SIZE a public constant field in the `HttpConfig` like so:

public final class HttpConfig {

    public static final int DEFAULT_MAX_HEADER_SIZE = 8192;


And then refer to it from `BackendService` like so:

    public static final int DEFAULT_RESPONSE_TIMEOUT_MILLIS = 1000;
    public static final int DEFAULT_MAX_HEADER_SIZE = HttpConfig.DEFAULT_MAX_HEADER_SIZE;

Then this would simplify the code and make the default value more obvious.

Reply:
Okay. @a-dlatorre this doesn't work because the BackendService class is in the api-module, and HttpConfig is in the client module. Therefore BackendService doesn't have access to the HttpConfig definitions.